src/object/schema_class_truncator.cpp

```diff
@@ -0,0 +1,634 @@
+    public:
+      using cons_predicate = std::function<bool (const SM_CLASS_CONSTRAINT &)>;
+      using saved_cons_predicate = std::function<bool (const SM_CONSTRAINT_INFO &)>;
+
+      int save_constraints_or_clear (cons_predicate pred);
+      int drop_saved_constraints (saved_cons_predicate pred);
+      int truncate_heap ();
+      int restore_constraints (saved_cons_predicate pred);
+      int reset_serials ();
+
+      class_truncate_context (const OID &class_oid);
+      class_truncate_context (class_truncate_context &&other);
+      ~class_truncate_context ();
+
+      class_truncate_context (const class_truncate_context &other) = delete;
+      class_truncate_context &operator= (const class_truncate_context &other) = delete;
+      class_truncate_context &operator= (const class_truncate_context &&other) = delete;
+
+    private:
+      MOP m_mop = NULL;
+      SM_CLASS *m_class = NULL;
+      SM_CONSTRAINT_INFO *m_unique_info = NULL;
+      SM_CONSTRAINT_INFO *m_fk_info = NULL;
+      SM_CONSTRAINT_INFO *m_index_info = NULL;
+  };
+
+  /*
+   * collect_trun_classes () - Collects OIDs of truncatable classes regarding the CASCADE option
+   *   return: NO_ERROR on success, non-zero for ERROR
+   *   class_mop(in):
+   *   is_cascade(in): whether to cascade TRUNCATE to FK-referring classes
+   */
+  int
+  class_truncator::collect_trun_classes (MOP class_mop, bool is_cascade)
+  {
+    int error = NO_ERROR;
+    SM_CLASS *class_ = NULL;
+    SM_CLASS_CONSTRAINT *pk_constraint = NULL;
+    SM_FOREIGN_KEY_INFO *fk_ref;
+    OID *fk_cls_oid;
+
+    error = au_fetch_class (class_mop, &class_, AU_FETCH_READ, DB_AUTH_ALTER);
+    if (error != NO_ERROR || class_ == NULL)
+      {
+	assert (er_errid () != NO_ERROR);
+	return er_errid ();
+      }
+
+    m_trun_classes.emplace (*ws_oid (class_mop));
+
+    pk_constraint = classobj_find_cons_primary_key (class_->constraints);
+    if (pk_constraint == NULL || classobj_is_pk_referred (class_mop, pk_constraint->fk_info, false, NULL) == false)
+      {
+	/* if no PK or FK-referred, it can be truncated */
+	return NO_ERROR;
+      }
+
+    /* Now, there is a PK, and are some FKs-referring to the PK */
+
+    if (!is_cascade)
+      {
+	er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TRUNCATE_PK_REFERRED, 1, pk_constraint->fk_info->name);
+	return ER_TRUNCATE_PK_REFERRED;
+      }
+
+    /* Find FK-child classes to cascade. */
+    for (fk_ref = pk_constraint->fk_info; fk_ref; fk_ref = fk_ref->next)
+      {
+	if (fk_ref->delete_action == SM_FOREIGN_KEY_CASCADE)
+	  {
+	    if (m_trun_classes.find (fk_ref->self_oid) != m_trun_classes.end ())
+	      {
+		continue;		/* already checked */
+	      }
+
+	    MOP fk_child_mop = ws_mop (&fk_ref->self_oid, NULL);
+	    if (fk_child_mop == NULL)
+	      {
+		assert (er_errid () != NO_ERROR);
+		return er_errid ();
+	      }
+
+	    int partition_type = DB_NOT_PARTITIONED_CLASS;
+	    error = sm_partitioned_class_type (fk_child_mop, &partition_type, NULL, NULL);
```

Comment:
I think you need to check the truncate target table when collect_trun_classes() is first called.

eg.) truncate table [target table: partitioned table];

It looks like checking only the fk-child tables. And I think it makes confuse to code reader even If it is not acceptable caused for an architectural reason. If so, I request you to leave a comment about it.

Reply:
If you mean the case in which the target table is  '**partition** table', it is already filtered out in a caller, `do_truncate()`. If it is what you talked about, yes, you're right, I'll add a comment for it.

Reply:
@wrlawodms Thanks. But, I reconsider after leaving this comment. And I understand the mechanism more precisely. There are no need changes and comments.

Reply:
OK!